### ISSUE TITLE
Request for training recipe of on-policy KD

### ISSUE THREAD

**Issue by JihwanEom** [NVIDIA-NeMo/RL#1527]
Hello! Thank you for the excellent work. I’m trying to **reproduce Table 1** from the post (https://github.com/NVIDIA-NeMo/RL/discussions/1445) and would be very grateful for any guidance. A few targeted questions:

1. Could you share the training recipe used for on-policy KD in the post? (Qwen3-32B teacher, Qwen3-4B student, how many gpus did you used?)

2. If it's sharable.. could you share training logs? I'd like to refer the training curves of losses/evaluation metrics to reproduce the experiment. It would be a great help. 

3. Could you tell some practical tips for on-policy kd training? for example ..
- The sensitivity for learning rate  
- KL loss type (mixed for revered, kl weight)
- The appropriate performance gap between student and teacher?
- Tricks to avoid model collapse

Sorry for many questions and thanks again for sharing the great work on open-source community!

Best regards,  
Jihwan

**Comment by zpqiu** [comment]
Thanks for your interest, @JihwanEom. Since some experiments in the blog were conducted on an early version of the PR, we need some time to organize our recipe before sharing it with you.

For part 3, I can give some quick answers based on our recent experiments:

- For lr, 2e-5 works well in our experiments, while 4e-5 performs worse at the beginning but aligns with 2e-5 after about 100 steps.
- Reverse KL works better.
<img width="2206" height="561" alt="Image" src="https://github.com/user-attachments/assets/edfb6951-97a0-465d-b786-def0afb9c018" />

- We haven't tried many student–teacher pairs, but for math tasks, using Qwen3-4B as the teacher can also improve the performance of 1.7B-Base.
- Top-k should be appropriate, at least 64. A larger batch size, such as 512, seems to make training more stable.

_But please note that all the tips come from our limited experimental observations, so they may not work in all scenarios._

**Comment by zpqiu** [comment]
Hi @JihwanEom , I has tested the following recipe.

```yaml
defaults: distillation_math.yaml
distillation:
  num_prompts_per_step: 512
  max_num_steps: 500
  val_batch_size: 512
  val_period: 20
loss_fn:
  kl_type: reverse
checkpointing:
  model_save_format: "torch_save"
  keep_top_k: 3
  checkpoint_dir: checkpoints/distillation-qwen3-32b-to-4b-base-long
policy:
  model_name: Qwen/Qwen3-4B-Base
  train_global_batch_size: 512
  max_total_sequence_length: 20480
  generation:
    vllm_cfg:
      tensor_parallel_size: 1
      gpu_memory_utilization: 0.7
teacher:
  model_name: Qwen/Qwen3-32B
  max_total_sequence_length: 20480
logger:
  log_dir: logs/distillation-qwen3-32b-to-4b-base-long
  wandb:
    project: nemo-rl
    name: distillation-qwen3-32b-to-4b-base-long
cluster:
  num_nodes: 2
```
The training curves are like:
<img width="1163" height="320" alt="Image" src="https://github.com/user-attachments/assets/9da3d192-45d0-4cff-95eb-9aa8afa9819d" />

At step 60, the score on AIME 2024 can reach 37.29, and the score on AIME 2025 can reach 32.71. The evaluation settings are:
```text
max_new_tokens=20480 temperature=0.6 top_p=1.0 top_k=-1 seed=42
metric=pass@1 num_tests_per_prompt=16
```

Therefore, the above configuration should roughly reproduce the results in the table of our blog, which can serve as a reference for you :)

**Comment by zpqiu** [comment]
Regarding Q4, yes, I don’t think there’s any difference between VLM and LLM in terms of on-policy distillation except for the input. Have you tried directly changing the model name to an VLM and modifying the multimodal dataset?

Regarding Q5, my understanding is that as long as Nemo-RL supports this vllm feature on the generation worker, on-policy distillation will naturally be supported as well. so we could ask @terrykong about his plans.

**Comment by zpqiu** [comment]
> I found that the response length will explode because the base model and instruct model have different EOS models.

The mismatch in EOS tokens does indeed cause the response to fail to stop correctly. In fact, in the third figure above, the truncation rate of the response is quite high (usually occurring because the model keeps repeating tokens after providing the correct answer). 

A possible solution is to add the logit assigned by the Instruct model to its own EOS token onto the student model's EOS token index before calculating the KL divergence, similar to the code block below:

```python
# EOS token ID correction: replace teacher EOS with student EOS if they differ
if (self.teacher_eos_token_id is not None and 
    self.student_eos_token_id is not None and 
    self.teacher_eos_token_id != self.student_eos_token_id):
    teacher_topk_indices = torch.where(
        teacher_topk_indices == self.teacher_eos_token_id,
        torch.tensor(self.student_eos_token_id, dtype=teacher_topk_indices.dtype, device=teacher_topk_indices.device),
        teacher_topk_indices
    )
```
